vllm/model_executor/models/llama_with_lora.py

```diff
@@ -0,0 +1,424 @@
+
+from vllm.model_executor.input_metadata import InputMetadata
+from vllm.model_executor.layers.activation import SiluAndMul
+from vllm.model_executor.layers.layernorm import RMSNorm
+from vllm.model_executor.layers.attention import PagedAttentionWithRoPE
+from vllm.model_executor.layers.sampler import Sampler
+from vllm.model_executor.weight_utils import (
+    load_tensor_parallel_weights, load_padded_tensor_parallel_vocab,
+    hf_model_weights_iterator)
+from vllm.model_executor.parallel_utils.parallel_state import (
+    get_tensor_model_parallel_rank, get_tensor_model_parallel_world_size)
+from vllm.model_executor.parallel_utils.tensor_parallel import (
+    VocabParallelEmbedding, ColumnParallelLinear, RowParallelLinear)
+from vllm.sequence import SamplerOutput
+
+KVCache = Tuple[torch.Tensor, torch.Tensor]
+
+
+class LlamaMLP(nn.Module):
+
+    def __init__(
+        self,
+        hidden_size: int,
+        intermediate_size: int,
+        hidden_act: str,
+    ):
+        super().__init__()
+        self.gate_up_proj = ColumnParallelLinear(hidden_size,
+                                                 2 * intermediate_size,
+                                                 bias=False,
+                                                 gather_output=False,
+                                                 perform_initialization=False)
+        self.down_proj = RowParallelLinear(intermediate_size,
+                                           hidden_size,
+                                           bias=False,
+                                           input_is_parallel=True,
+                                           perform_initialization=False)
+        if hidden_act != "silu":
+            raise ValueError(f"Unsupported activation: {hidden_act}. "
+                             "Only silu is supported for now.")
+        self.act_fn = SiluAndMul()
+
+    def forward(self, x):
+        gate_up, _ = self.gate_up_proj(x)
+        x = self.act_fn(gate_up)
+        x, _ = self.down_proj(x)
+        return x
+
+
+class LinearLoRALayer(nn.Module):
+    def __init__(self, hidden_dim, rank, dropout, scaling):
+        super().__init__()
+        self.lora_A = nn.Linear(hidden_dim, rank, bias=False)
+        self.lora_B = nn.Linear(rank, hidden_dim, bias=False)
+        self.scaling = scaling
+        if dropout > 0.0:
+            self.dropout = nn.Dropout(dropout)
+        else:
+            self.dropout = nn.Identity()
+
+    @classmethod
+    def from_state_dict(cls, state_dict):
+        expected_keys = set(['loras', 'dropout', 'scaling'])
+        if state_dict.keys() != expected_keys:
+            raise ValueError(f"Expected keys {expected_keys}, got {state_dict.keys()}")
+
+        rank, hidden_dim = state_dict['loras']['lora_A.weight'].shape
+        dropout = state_dict['dropout']
+        scaling = state_dict['scaling']
+        lora_layer = cls(hidden_dim=hidden_dim, rank=rank, dropout=dropout, scaling=scaling)
```

Comment:
this may be slow

Reply:
Ooh curious, what part of this is slow? Happy to improve it, I just don't know what to improve

Reply:
sorry this is supposed to be just about L103. cls() means  LinearLoRALayer.__init__ which can call torch.nn.init.uniform_ on every parameter